File: src/resolvers/model_resolver.py
"""
model_resolver.py — Canonical model resolver for AAA federation.
Reads strictly from /root/.config/federation-models.json.
Zero hardcoded model strings. Drift impossible by construction.

Usage:
    from model_resolver import resolve_agent_model, resolve_tier_model, list_agents

    cfg = resolve_agent_model("opencode")
    # → {"provider": "deepseek", "model": "deepseek/deepseek-v4-pro", ...}

    model = resolve_tier_model("heavy")
    # → "deepseek/deepseek-v4-pro"

Forged: 2026-07-24 by FORGE (000Ω). DITEMPA BUKAN DIBERI.
"""

import json
from pathlib import Path
from typing import Optional, Dict, List

CANONICAL_PATH = Path("/root/.config/federation-models.json")


def _load_registry() -> dict:
    """Load canonical registry. Raises FileNotFoundError if missing."""
    if not CANONICAL_PATH.exists():
        raise FileNotFoundError(
            f"Canonical model registry not found at {CANONICAL_PATH}. Run: forge --option-a to create it."
        )
    with open(CANONICAL_PATH) as f:
        return json.load(f)


def resolve_agent_model(agent_id: str) -> dict:
    """
    Resolve an agent's model configuration from canonical registry.

    Args:
        agent_id: Agent identifier (e.g. 'opencode', 'kimi-code', 'hermes')

    Returns:
        dict with keys: provider, model, fallback_chain, timeout_ms, cost_budget_daily_usd

    Raises:
        KeyError: Agent not found in registry
        FileNotFoundError: Canonical registry missing
    """
    registry = _load_registry()
    agents = registry.get("agents", {})

    if agent_id not in agents:
        raise KeyError(f"Agent '{agent_id}' not found in canonical registry. Available agents: {list(agents.keys())}")

    agent_cfg = agents[agent_id]
    provider_id = agent_cfg.get("primary_provider", "unknown")
    providers = registry.get("providers", {})
    provider_cfg = providers.get(provider_id, {})

    return {
        "agent_id": agent_id,
        "agent_name": agent_cfg.get("name", agent_id),
        "provider": provider_id,
        "provider_name": provider_cfg.get("name", provider_id),
        "model": agent_cfg["primary_model"],
        "fallback_chain": agent_cfg.get("fallback_chain", []),
        "timeout_ms": agent_cfg.get("timeout_ms", 30000),
        "cost_budget_daily_usd": agent_cfg.get("cost_budget_daily_usd", 0),
        "api_key_ref": provider_cfg.get("api_key_ref", ""),
        "endpoint": provider_cfg.get("endpoint", ""),
        "status": agent_cfg.get("status", "UNKNOWN"),
    }


def resolve_tier_model(tier: str) -> str:
    """
    Resolve the model for a given usage tier.

    Args:
        tier: One of 'bulk', 'default', 'heavy', 'apex'

    Returns:
        Model string (e.g. 'deepseek/deepseek-v4-pro')
    """
    registry = _load_registry()
    tiers = registry.get("tiers", {})

    if tier not in tiers:
        raise KeyError(f"Tier '{tier}' not found. Available: {list(tiers.keys())}")

    return tiers[tier]["model"]


def resolve_fallback(agent_id: str, failed_model: str) -> Optional[str]:
    """
    Given a failed model, return the next fallback for an agent.

    Args:
        agent_id: Agent identifier
        failed_model: The model that failed (to skip past)

    Returns:
        Next model in fallback chain, or None if exhausted
    """
    cfg = resolve_agent_model(agent_id)
    chain = cfg["fallback_chain"]

    # Find current position in chain
    for i, m in enumerate(chain):
        if m == failed_model or failed_model in m:
            if i + 1 < len(chain):
                return chain[i + 1]
            break

    # If not found, return first fallback
    return chain[0] if chain else None


def list_agents(status_filter: Optional[str] = None) -> List[dict]:
    """
    List all agents with their model assignments.

    Args:
        status_filter: Optional filter by status (ACTIVE, STANDBY, etc.)

    Returns:
        List of agent summaries
    """
    registry = _load_registry()
    agents = []

    for aid, cfg in registry.get("agents", {}).items():
        status = cfg.get("status", "UNKNOWN")
        if status_filter and status != status_filter:
            continue
        agents.append(
            {
                "agent_id": aid,
                "name": cfg.get("name", aid),
                "model": cfg["primary_model"],
                "provider": cfg.get("primary_provider", "unknown"),
                "status": status,
            }
        )

    return sorted(agents, key=lambda x: x["agent_id"])


def list_providers() -> List[dict]:
    """List all providers with status and balance."""
    registry = _load_registry()
    providers = []

    for pid, cfg in registry.get("providers", {}).items():
        providers.append(
            {
                "provider_id": pid,
                "name": cfg.get("name", pid),
                "status": cfg.get("status", "UNKNOWN"),
                "balance_usd": cfg.get("balance_usd"),
                "jurisdiction": cfg.get("jurisdiction", "unknown"),
                "billing": cfg.get("billing", "unknown"),
            }
        )

    return sorted(providers, key=lambda x: x["provider_id"])


# CLI entry point
if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        print("Usage: python model_resolver.py <agent_id>")
        print("       python model_resolver.py --list")
        print("       python model_resolver.py --tier <tier>")
        print("       python model_resolver.py --providers")
        sys.exit(1)

    if sys.argv[1] == "--list":
        for a in list_agents():
            print(f"{a['agent_id']:15s} → {a['model']:40s} ({a['status']})")

    elif sys.argv[1] == "--tier" and len(sys.argv) > 2:
        model = resolve_tier_model(sys.argv[2])
        print(f"tier={sys.argv[2]} → {model}")

    elif sys.argv[1] == "--providers":
        for p in list_providers():
            bal = f"${p['balance_usd']}" if p["balance_usd"] is not None else "N/A"
            print(f"{p['provider_id']:20s} {p['status']:12s} {bal:>8s}  {p['jurisdiction']}")

    else:
        agent_id = sys.argv[1]
        cfg = resolve_agent_model(agent_id)
        print(json.dumps(cfg, indent=2, default=str))

File: src/resolvers/model_resolver.ts
/**
 * model_resolver.ts — Canonical model resolver for AAA federation (TypeScript).
 * Reads from AGENT_MODEL_MAP.json via symlink at /root/.config/federation-models.json.
 * Schema: providers[] + agents[] + models[] + fallback_chains[].
 *
 * Usage:
 *   import { resolve, validate, listAgents } from './model_resolver';
 *   const cfg = resolve('opencode');
 *   // → { agentId: 'opencode', model: 'deepseek/deepseek-v4-pro', ... }
 *
 * Forged: 2026-07-24 by FORGE (000Ω). DITEMPA BUKAN DIBERI.
 */

import * as fs from 'fs';

const CANONICAL = '/root/.config/federation-models.json';

// ── Types ──

interface ProviderRecord {
  provider_id: string;
  provider_name?: string;
  endpoint_url?: string;
  api_key_ref?: string;
  status?: string;
  balance_usd?: number | null;
  jurisdiction?: string;
  billing_model?: string;
  cloud_act_exposed?: boolean;
  note?: string;
}

interface FallbackEntry {
  model_key: string;
  priority: number;
  condition: string;
}

interface AgentRecord {
  agent_id: string;
  agent_name?: string;
  primary_model: string;
  primary_provider: string;
  fallback_chain: FallbackEntry[];
  timeout_ms?: number;
  cost_budget_daily_usd?: number;
  status?: string;
}

interface ModelRecord {
  model_key: string;
  provider_ref: string;
  status?: string;
}

interface FederationRegistry {
  providers: ProviderRecord[];
  agents: AgentRecord[];
  models: ModelRecord[];
  fallback_chains?: unknown[];
  routing_rules?: unknown[];
}

interface ResolvedAgent {
  agentId: string;
  agentName: string;
  model: string;
  provider: string;
  providerName: string;
  fallbacks: string[];
  endpoint: string;
  apiKeyRef: string;
  status: string;
  timeoutMs: number;
  costDailyUsd: number;
}

interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
  providers: number;
  agents: number;
}

// ── Cache ──

let _cache: FederationRegistry | null = null;
let _cacheTime = 0;
const CACHE_TTL = 60_000;

function load(): FederationRegistry {
  const now = Date.now();
  if (_cache && now - _cacheTime < CACHE_TTL) return _cache;
  if (!fs.existsSync(CANONICAL)) {
    throw new Error(`Registry missing: ${CANONICAL}`);
  }
  _cache = JSON.parse(fs.readFileSync(CANONICAL, 'utf-8')) as FederationRegistry;
  _cacheTime = now;
  return _cache;
}

export function reload(): void {
  _cache = null;
}

// ── Index helpers ──

function providerMap(reg: FederationRegistry): Map<string, ProviderRecord> {
  const m = new Map<string, ProviderRecord>();
  for (const p of reg.providers) m.set(p.provider_id, p);
  return m;
}

function agentMap(reg: FederationRegistry): Map<string, AgentRecord> {
  const m = new Map<string, AgentRecord>();
  for (const a of reg.agents) m.set(a.agent_id, a);
  return m;
}

// ── Public API ──

export function resolve(agentId: string): ResolvedAgent {
  const reg = load();
  const agents = agentMap(reg);
  const providers = providerMap(reg);

  const a = agents.get(agentId);
  if (!a) throw new Error(`Agent '${agentId}' not found. Available: ${[...agents.keys()].join(', ')}`);

  const pid = a.primary_provider || 'unknown';
  const p = providers.get(pid);

  const fallbacks: string[] = (a.fallback_chain || [])
    .filter((f): f is FallbackEntry => typeof f === 'object' && 'model_key' in f)
    .map(f => f.model_key);

  return {
    agentId,
    agentName: a.agent_name || agentId,
    model: a.primary_model,
    provider: pid,
    providerName: p?.provider_name || pid,
    fallbacks,
    endpoint: p?.endpoint_url || '',
    apiKeyRef: p?.api_key_ref || '',
    status: a.status || p?.status || 'UNKNOWN',
    timeoutMs: a.timeout_ms || 30000,
    costDailyUsd: a.cost_budget_daily_usd || 0,
  };
}

export function resolveFallback(agentId: string, afterModel: string): string | null {
  const cfg = resolve(agentId);
  const chain = cfg.fallbacks;
  for (let i = 0; i < chain.length; i++) {
    if (chain[i].includes(afterModel) || afterModel.includes(chain[i])) {
      return i + 1 < chain.length ? chain[i + 1] : null;
    }
  }
  return chain.length > 0 ? chain[0] : null;
}

export function validate(): ValidationResult {
  const reg = load();
  const errors: string[] = [];
  const warnings: string[] = [];
  const providers = providerMap(reg);
  const agents = agentMap(reg);

  // Build known models set
  const knownModels = new Set<string>();
  for (const m of reg.models || []) {
    if (m.model_key) knownModels.add(m.model_key);
  }

  for (const [aid, a] of agents) {
    if (!a.primary_model) errors.push(`${aid}: missing primary_model`);
    if (!providers.has(a.primary_provider)) {
      errors.push(`${aid}: provider '${a.primary_provider}' not in registry`);
    }
    for (const fb of a.fallback_chain || []) {
      const mk = fb.model_key || '';
      if (mk && !knownModels.has(mk)) {
        const found = [...knownModels].some(km => mk.includes(km) || km.includes(mk));
        if (!found) warnings.push(`${aid}: fallback model '${mk}' not in registry models catalog`);
      }
    }
  }

  for (const [pid, p] of providers) {
    if (p.status === 'DEPRECATING' || p.status === 'RATE_LIMITED') {
      warnings.push(`${pid}: ${p.status} — agents using this may fail`);
    }
    if (p.cloud_act_exposed) {
      warnings.push(`${pid}: US jurisdiction — not for sovereign data`);
    }
  }

  return { valid: errors.length === 0, errors, warnings, providers: providers.size, agents: agents.size };
}

export function listAgents(): { agentId: string; name: string; model: string; provider: string; status: string }[] {
  const reg = load();
  return reg.agents
    .map(a => ({
      agentId: a.agent_id,
      name: a.agent_name || a.agent_id,
      model: a.primary_model,
      provider: a.primary_provider || 'unknown',
      status: a.status || '?',
    }))
    .sort((a, b) => a.agentId.localeCompare(b.agentId));
}

export function listProviders(): { providerId: string; name: string; status: string; balanceUsd: number | null; jurisdiction: string }[] {
  const reg = load();
  return reg.providers
    .map(p => ({
      providerId: p.provider_id,
      name: p.provider_name || p.provider_id,
      status: p.status || '?',
      balanceUsd: p.balance_usd ?? null,
      jurisdiction: p.jurisdiction || '?',
    }))
    .sort((a, b) => a.providerId.localeCompare(b.providerId));
}
